safehttp/form_test.go

```diff
@@ -12,1135 +12,476 @@
+			values:      []string{"1", "abc", "-1"},
 			placeholder: &[]uint64{},
-			want:        &[]uint64{8, 9, 10},
 		},
 		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=margeritta&pizza=diavola&pizza=calzone"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
-			placeholder: &[]string{},
-			want:        &[]string{"margeritta", "diavola", "calzone"},
-		},
-		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=1.3&pizza=8.9&pizza=-4.1"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
+			name:        "Float64",
+			values:      []string{"1.3", "abc", "-4.1"},
 			placeholder: &[]float64{},
-			want:        &[]float64{1.3, 8.9, -4.1},
 		},
 		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=true&pizza=false&pizza=T"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
+			name:        "Bool",
+			values:      []string{"t", "abc", "TRUE"},
 			placeholder: &[]bool{},
-			want:        &[]bool{true, false, true},
 		},
 	}
 
-	for _, tc := range tests {
-		mux := safehttp.NewServeMux(testDispatcher{}, "foo.com")
-		mux.Handle("/", safehttp.MethodPost, safehttp.HandlerFunc(func(rw safehttp.ResponseWriter, ir *safehttp.IncomingRequest) safehttp.Result {
-			form, err := ir.PostForm()
-			if err != nil {
-				t.Fatalf(`ir.PostForm: got err %v, want nil`, err)
-			}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			f := Form{values: map[string][]string{"x": tt.values}}
 
-			form.Slice("pizza", tc.placeholder)
-			if err := form.Err(); err != nil {
-				t.Errorf(`form.Err: got err %v, want nil`, err)
-			}
-
-			if diff := cmp.Diff(tc.want, tc.placeholder); diff != "" {
-				t.Errorf("form.Slice: diff (-want +got): \n%v", diff)
-			}
+			f.Slice("x", tt.placeholder)
 
-			return safehttp.Result{}
-		}))
+			// TODO: add a check here that tt.placeholder is nil. I (grenfeldt@)
+			// can't come up with a way of testing this in a table test.
 
-		rec := newResponseRecorder(&strings.Builder{})
-		mux.ServeHTTP(rec, tc.req)
-
-		if respStatus, want := rec.status, 200; respStatus != want {
-			t.Errorf("response status: got %v, want %v", respStatus, want)
-		}
+			if err := f.Err(); err == nil {
+				t.Error("f.Err() got: nil want: error")
+			}
+		})
 	}
-
 }
 
-func TestFormValidSliceMultipart(t *testing.T) {
-	multipartBody := template.Must(template.New("multipart").Parse(
-		"{{ range . -}}" +
-			"--123\r\n" +
-			"Content-Disposition: form-data; name=\"pizza\"\r\n" +
-			"\r\n" +
-			"{{ . }}\r\n" +
-			"{{ end }}" +
-			"--123--\r\n"))
+func TestFormSliceInvalidPointerType(t *testing.T) {
+	f := Form{values: map[string][]string{"x": {"x"}}}
+	f.Slice("x", &[]int16{})
```

Comment:
Make sure the result is nil here.

Reply:
`f.Slice` doesn't set pointers of unknown type to nil.

Reply:
Ah right, sorry, my bad.